From 47048edeb09e73af44bbe958b90ecfe96265b328 Mon Sep 17 00:00:00 2001
From: Dean Sanner <dsanner@us.ibm.com>
Date: Fri, 25 Jan 2019 10:24:36 -0600
Subject: [PATCH v1] Support direct load of POWERVM from BOOTKERNEL partition

 PowerVM doesn't fit in the PAYLOAD section of PNOR.  To
 enable a common PNOR layout instead put PowerVM into
 BOOTKERNEL partition, and if PHYP payload, load BOOTKERNEL
 instead of PAYLOAD PNOR partition

Change-Id: Id19f3ce708b537a19256c22aab3cc1b3e83d4ab1
---
 src/include/usr/pnor/pnor_const.H               |  1 +
 src/usr/isteps/istep20/call_host_load_payload.C | 10 ++++++++--
 src/usr/pnor/pnor_utils.C                       |  2 ++
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/include/usr/pnor/pnor_const.H b/src/include/usr/pnor/pnor_const.H
index b68f5dd..4506e89 100644
--- a/src/include/usr/pnor/pnor_const.H
+++ b/src/include/usr/pnor/pnor_const.H
@@ -77,6 +77,7 @@ enum SectionId
     HDAT,           /**< HDAT data */
     EECACHE,
     UVBWLIST,       /**< Ultravisor XSCOM white/blacklist binary */
+    BOOTKERNEL,     /**< Bootkernel -- HB uses for PHYP */
 #endif
     NUM_SECTIONS,   /**< Number of defined sections */
 
diff --git a/src/usr/isteps/istep20/call_host_load_payload.C b/src/usr/isteps/istep20/call_host_load_payload.C
index 6251eaf..808d87d 100644
--- a/src/usr/isteps/istep20/call_host_load_payload.C
+++ b/src/usr/isteps/istep20/call_host_load_payload.C
@@ -5,7 +5,7 @@
 /*                                                                        */
 /* OpenPOWER HostBoot Project                                             */
 /*                                                                        */
-/* Contributors Listed Below - COPYRIGHT 2016,2017                        */
+/* Contributors Listed Below - COPYRIGHT 2016,2019                        */
 /* [+] International Business Machines Corp.                              */
 /*                                                                        */
 /*                                                                        */
@@ -120,7 +120,13 @@ void* call_host_load_payload (void *io_pArgs)
         // Load payload data in PHYP mode or in Sapphire mode
         if(is_sapphire_load() || is_phyp_load())
         {
-            l_err = load_pnor_section( PNOR::PAYLOAD, payloadBase );
+            PNOR::SectionId l_secID = PNOR::PAYLOAD;
+            if (is_phyp_load())
+            {
+                l_secID = PNOR::BOOTKERNEL;
+            }
+
+            l_err = load_pnor_section( l_secID, payloadBase );
             if ( l_err )
             {
                 TRACFCOMP(ISTEPS_TRACE::g_trac_isteps_trace,
diff --git a/src/usr/pnor/pnor_utils.C b/src/usr/pnor/pnor_utils.C
index 95989cd..f027707 100644
--- a/src/usr/pnor/pnor_utils.C
+++ b/src/usr/pnor/pnor_utils.C
@@ -369,6 +369,7 @@ bool PNOR::isEnforcedSecureSection(const uint32_t i_section)
                i_section == HB_DATA ||
                i_section == SBE_IPL ||
                i_section == PAYLOAD ||
+               i_section == BOOTKERNEL ||
                i_section == SBKT ||
                i_section == OCC ||
                i_section == HCODE ||
@@ -452,6 +453,7 @@ const char * PNOR::SectionIdToString( uint32_t i_secIdIndex )
         "HDAT",        /**< PNOR::HDAT           : Hdat Data */
         "EECACHE",     /**< PNOR::EECACHE        : Cached data from various EEPROMs */
         "UVBWLIST",    /**< PNOR::UVBWLIST       : Ultravisor XSCOM white/blacklist */
+        "BOOTKERNEL",  /**< PNOR::BOOTKERNEL     : OPAL == petitboot,PHYP == PowerVM */
 #endif
     };
 
-- 
1.8.2.2

